R/hub-connection.R

```diff
@@ -0,0 +1,178 @@
+#' hub_path <- system.file("hub_1", package = "hubUtils")
+#' connect_hub(hub_path)
+#' # Scenario Hub example
+#' scnr_path <- system.file("scnr_hub_1", package = "hubUtils")
+#' connect_hub(hub_path)
+connect_hub <- function(hub_path, hubmeta_path = NULL,
+                        hubmeta_format = c("json", "yaml", "yml")) {
+
+    hubmeta_format <- rlang::arg_match(hubmeta_format)
+
+    if (is.null(hubmeta_path)) {
+        hubmeta_path <- fs::path(
+            hub_path,
+            "hubmeta",
+            ext = hubmeta_format
+            )
+    }
+
+    hubmeta <- read_hubmeta(hubmeta_path)
+
+    hubmeta[["hubmeta_path"]] <- hubmeta_path
+    hubmeta[["hub_path"]] <- hub_path
+
+    new_hub_connection(hubmeta)
+
+}
+
+# INTERNAL ----
+
+#' Create new hub connection S3 class object
+#'
+#' Internal function for create a new hub connection S3 class object
+#' @param x A metadata list generated by [read_hubmeta()].
+#'
+#' @return A hub connection S3 class object.
+#' @noRd
+new_hub_connection <- function(x) {
+    stopifnot(is.list(x))
+
+    x <- assing_hc_attrs(x)
+
+    class(x) <- "hub_connection"
+
+    x
+}
+
+
+#' Assign high level hub metadata as attributes.
+#'
+#' @param x A metadata list generated by [read_hubmeta()].
+#'
+#' @return
+#'   A metadata list with `task_ids` names and `round_ids` assigned to
+#'   attributes `task_id_names` and `round_ids` respectively.
+#'
+#'   Where the metadata list contains round specific metadata
+#'   elements, a named list containing `task_ids` names per round is assigned to
+#'   `task_id_names` and a character vector of the names of elements containing
+#'   round specific metadata is assigned to `round_ids`.
+#'
+#'   Where metadata are shared between rounds
+#'   (indicated by the name of the first element of the metadata list being
+#'   `"round_id_from_variable"`), a single character vector of task_ids names
+#'   is assigned to attribute `task_id_names` and the value of element
+#'   `"round_id_from_variable"` is assigned to attribute `round_ids`.
+#'
+#'   Where a `hubmeta_path` and/or `hub_path` elements are present in the metadata
+#'   list, they are removed from the list and assigned to attributes `hubmeta_path`
+#'   `hub_path` respectively.
+#' @noRd
+assing_hc_attrs <- function(x) {
+
+    # if assigned, move hubmeta_path & hub_path to attributes to maintain
+    # original hubmeta structure
+    if(!is.null(x$hubmeta_path)){
+        attr(x, "hubmeta_path") <- x$hubmeta_path
+        x$hubmeta_path <- NULL
+    }
+
+    if(!is.null(x$hub_path)){
+        attr(x, "hub_path") <- x$hub_path
+        x$hub_path <- NULL
+    }
+
+    if (names(x)[1] == "round_id_from_variable") {
+
+        task_id_names <- purrr::pluck(
+            x,
+            "round_id_from_variable",
+            "model_tasks",
+            1,
+            "task_ids"
+        ) %>%
+            names()
+
+        round_ids <- x[["round_id_from_variable"]][["round_id_variable"]]
```

Comment:
I'm missing something about how this works. e.g., I was thinking that we'd need to do something like `x[["round_id_from_variable"]]...[[x[["round_id_variable"]]]]`?

Reply:
Here we're extracting the variable specified in `round_id_variable` and assigning to the `round_ids` attribute. All metadata in non round varying metadata are contained in top level `round_id_from_variable` and then we select the `round_id_variable`.

In all honesty, at the minute, this is largely ignored when validating non round varying hubs (as descussed above and in #22) but post #22 implementation, the idea it is it will play a role in validation.


``` r
library(hubUtils)
x <- read_hubmeta(system.file("hub_1/hubmeta.json",
                              package = "hubUtils"))
str(x)
#> List of 1
#>  $ round_id_from_variable:List of 3
#>   ..$ model_tasks      :List of 1
#>   .. ..$ :List of 2
#>   .. .. ..$ task_ids    :List of 3
#>   .. .. .. ..$ origin_date:List of 2
#>   .. .. .. .. ..$ required: NULL
#>   .. .. .. .. ..$ optional: chr [1:24] "2022-01-08" "2022-01-15" "2022-01-22" "2022-01-29" ...
#>   .. .. .. ..$ location   :List of 2
#>   .. .. .. .. ..$ required: NULL
#>   .. .. .. .. ..$ optional: chr [1:54] "01" "02" "04" "05" ...
#>   .. .. .. ..$ horizon    :List of 2
#>   .. .. .. .. ..$ required: NULL
#>   .. .. .. .. ..$ optional: int [1:4] 1 2 3 4
#>   .. .. ..$ output_types:List of 2
#>   .. .. .. ..$ mean    :List of 2
#>   .. .. .. .. ..$ type_id:List of 2
#>   .. .. .. .. .. ..$ required: NULL
#>   .. .. .. .. .. ..$ optional: logi NA
#>   .. .. .. .. ..$ value  :List of 2
#>   .. .. .. .. .. ..$ type   : chr "integer"
#>   .. .. .. .. .. ..$ minimum: int 0
#>   .. .. .. ..$ quantile:List of 2
#>   .. .. .. .. ..$ type_id:List of 2
#>   .. .. .. .. .. ..$ required: NULL
#>   .. .. .. .. .. ..$ optional: num [1:23] 0.01 0.025 0.05 0.1 0.15 0.2 0.25 0.3 0.35 0.4 ...
#>   .. .. .. .. ..$ value  :List of 2
#>   .. .. .. .. .. ..$ type   : chr "integer"
#>   .. .. .. .. .. ..$ minimum: int 0
#>   ..$ round_id_variable: chr "origin_date"
#>   ..$ submissions_due  :List of 3
#>   .. ..$ relative_to: chr "origin_date"
#>   .. ..$ start      : int -4
#>   .. ..$ end        : int 2
x[["round_id_from_variable"]][["round_id_variable"]]
#> [1] "origin_date"
```

<sup>Created on 2022-10-24 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
OK, thanks.  Noting that this may end up getting re-worked depending on where we head with the discussion [here](https://github.com/reichlab/hub-infrastructure-experiments/pull/4)